tensorboard/notebook.py

```diff
@@ -0,0 +1,336 @@
+        return "\n<could not read %s>" % name
+      else:
+        return "\nContents of %s:\n%s" % (name, value.strip())
+    message = (
+        "ERROR: Failed to launch TensorBoard (exited with %d).%s%s" %
+        (
+            start_result.exit_code,
+            format_stream("stderr", start_result.stderr),
+            format_stream("stdout", start_result.stdout),
+        )
+    )
+    print_or_update(message)
+
+  elif isinstance(start_result, manager.StartTimedOut):
+    message = (
+        "ERROR: Timed out waiting for TensorBoard to start. "
+        "It may still be running as pid %d."
+        % start_result.pid
+    )
+    print_or_update(message)
+
+  else:
+    raise TypeError(
+        "Unexpected result from `manager.start`: %r.\n"
+        "This is a TensorBoard bug; please report it."
+        % start_result
+    )
+
+
+def _time_delta_from_info(info):
+  """Format the elapsed time for the given TensorboardInfo.
+
+  Args:
+    info: A TensorboardInfo value.
+
+  Returns:
+    A human-readable string describing the time since the server
+    described by `info` started: e.g., "2 days, 0:48:58".
+  """
+  now = datetime.datetime.now()
+  then = info.start_time
+  return str(now.replace(microsecond=0) - then.replace(microsecond=0))
+
+
+def _data_source_from_info(info):
+  """Format the data location for the given TensorboardInfo.
+
+  Args:
+    info: A TensorboardInfo value.
+
+  Returns:
+    A human-readable string describing the logdir or database connection
+    used by the server: e.g., "logdir /tmp/logs".
+  """
+  if info.db:
+    return "db %s" % info.db
+  else:
+    return "logdir %s" % info.logdir
+
+
+def display(port=None, height=None):
+  """Display a TensorBoard instance already running on this machine.
+
+  Args:
+    port: The port on which the TensorBoard server is listening, as an
+      `int`, or `None` to automatically select the most recently
+      launched TensorBoard.
+    height: The height of the frame into which to render the TensorBoard
+      UI, as an `int` number of pixels, or `None` to use a default value.
+  """
+  _display(port=port, height=height, print_message=True, display_handle=None)
+
+
+def _display(port=None, height=None, print_message=False, display_handle=None):
+  """Internal version of `display`.
+
+  Args:
+    port: As with `display`.
+    height: As with `display`.
+    print_message: True to print which TensorBoard instance was selected
+      for display (if applicable), or False otherwise.
+    display_handle: If not None, an IPython display handle into which to
+      render TensorBoard.
+  """
+  if height is None:
+    height = 600
+
+  if port is None:
+    infos = manager.get_all()
+    if not infos:
+      raise ValueError("Can't display TensorBoard: no known instances running.")
+    else:
+      info = max(manager.get_all(), key=lambda x: x.start_time)
+      port = info.port
+  else:
+    infos = [i for i in manager.get_all() if i.port == port]
```

Comment:
Seems like this would be better if it would be provided directly by the manager API, i.e.:

manager.get_info_by_port(port)

The start_time is an implementation detail that should be interpreted here IMO.

Reply:
> Seems like this would be better if it would be provided directly by
> the manager API, i.e.:
> 
> manager.get_info_by_port(port)

That was actually the case in the original sketch of the `manager` API,
which was cut (and moved here) at Nick’s suggestion.

One thing to note is that in implementing a general `get_info_by_port`
it’s not obvious whether the caller wants an approximate response
(trusting the result of `get_all()`) or a response that has actually
been liveness-checked for accuracy. Leaving this implementation here
enables us to do what’s locally appropriate without worrying about the
general case.

> The start_time is an implementation detail that should be interpreted
> here IMO.

Not sure what this means?

Reply:
Ok, I agree to defer on this though I think it's less desirable.